File: api/Controllers/PagesController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using ProjectManagementAPI.DTOs;
using ProjectManagementAPI.Filters;
using ProjectManagementAPI.Helpers;
using ProjectManagementAPI.Services.Abstractions;

namespace ProjectManagementAPI.Controllers;

// ═══════════════════════════════════════════════════════════════════════════
//  Pages
//
//  ทุก endpoint ผูกกับ workspace — [RequireWorkspace] อยู่ที่ระดับคลาส
// ═══════════════════════════════════════════════════════════════════════════
[ApiController]
[Route("api/v1/pages")]
[Authorize]
[RequireWorkspace]
public class PagesController(IPageTreeService tree) : ControllerBase
{
    /// <summary>ทั้ง tree สำหรับ sidebar (กรองตามสิทธิ์แล้ว)</summary>
    [HttpGet]
    public async Task<IActionResult> GetTree(CancellationToken ct)
        => (await tree.GetTreeAsync(ct)).ToActionResult();

    [HttpGet("{pageId:guid}")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> Get(Guid pageId, CancellationToken ct)
        => (await tree.GetAsync(pageId, ct)).ToActionResult();

    [HttpPost]
    [ProducesResponseType(StatusCodes.Status201Created)]
    public async Task<IActionResult> Create(CreatePageRequest request, CancellationToken ct)
    {
        var result = await tree.CreateAsync(request, ct);

        return result.IsSuccess
            ? result.ToCreatedResult($"/api/v1/pages/{result.Value.Id}")
            : result.ToActionResult();
    }

    [HttpPatch("{pageId:guid}")]
    public async Task<IActionResult> Update(
        Guid pageId, UpdatePageRequest request, CancellationToken ct)
        => (await tree.UpdateAsync(pageId, request, ct)).ToActionResult();

    /// <summary>ย้ายหน้าพร้อมลูกหลานทั้งหมด</summary>
    [HttpPost("{pageId:guid}/move")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<IActionResult> Move(
        Guid pageId, MovePageRequest request, CancellationToken ct)
        => (await tree.MoveAsync(pageId, request, ct)).ToActionResult();

    /// <summary>ย้ายไปถังขยะ (soft delete) พร้อมลูกหลาน</summary>
    [HttpDelete("{pageId:guid}")]
    public async Task<IActionResult> Delete(Guid pageId, CancellationToken ct)
        => (await tree.DeleteAsync(pageId, ct)).ToActionResult();

    [HttpPost("{pageId:guid}/restore")]
    public async Task<IActionResult> Restore(Guid pageId, CancellationToken ct)
        => (await tree.RestoreAsync(pageId, ct)).ToActionResult();

    /// <summary>ลบถาวร — owner/admin เท่านั้น ย้อนกลับไม่ได้</summary>
    [HttpDelete("{pageId:guid}/purge")]
    public async Task<IActionResult> Purge(Guid pageId, CancellationToken ct)
        => (await tree.PurgeAsync(pageId, ct)).ToActionResult();

    [HttpGet("trash")]
    public async Task<IActionResult> GetTrash(CancellationToken ct)
        => (await tree.GetTrashAsync(ct)).ToActionResult();

    // ═══════════════════════════════════════════════════════════════════════
    //  Maintenance
    //
    //  ancestor_ids และ access_root_id เป็นค่า denormalise ที่เพี้ยนได้
    //  endpoint สองตัวนี้มีไว้ตรวจและซ่อม ควรตั้ง cron เรียก /consistency
    //  ทุกวันในช่วงเดือนแรก ๆ แล้ว alert เมื่อไม่เป็นศูนย์
    // ═══════════════════════════════════════════════════════════════════════

    [HttpGet("maintenance/consistency")]
    public async Task<IActionResult> CheckConsistency(CancellationToken ct)
        => (await tree.CheckConsistencyAsync(ct)).ToActionResult();

    [HttpPost("maintenance/repair")]
    public async Task<IActionResult> Repair(CancellationToken ct)
        => (await tree.RepairAsync(ct)).ToActionResult();
}

File: server/src/pages/pages.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseUUIDPipe,
  Patch,
  Post,
} from '@nestjs/common';
import { ApiOperation, ApiTags } from '@nestjs/swagger';

import { PageTreeService } from './page-tree.service.js';
import {
  type CreatePageInput,
  createPageSchema,
  type MovePageInput,
  movePageSchema,
  type UpdatePageInput,
  updatePageSchema,
} from './pages.schema.js';
import { unwrap } from '../common/api-response.js';
import { RequireWorkspace } from '../common/route-metadata.js';
import { zodBody } from '../common/zod-body.pipe.js';

// ═══════════════════════════════════════════════════════════════════════════
//  Pages — ทุก endpoint ผูกกับ workspace
//
//  ⚠️ ลำดับของ route สำคัญ: 'trash' และ 'maintenance/*' ต้องประกาศ "ก่อน"
//     ':pageId' ไม่งั้น Express จะจับ /pages/trash เข้า handler ของ :pageId
//     แล้ว ParseUUIDPipe จะตอบ 400 แทนที่จะได้ถังขยะ
// ═══════════════════════════════════════════════════════════════════════════
@ApiTags('pages')
@Controller('pages')
@RequireWorkspace()
export class PagesController {
  constructor(private readonly tree: PageTreeService) {}

  @Get()
  @ApiOperation({ summary: 'ทั้ง tree สำหรับ sidebar (กรองตามสิทธิ์แล้ว)' })
  async getTree() {
    return unwrap(await this.tree.getTree());
  }

  @Get('trash')
  async getTrash() {
    return unwrap(await this.tree.getTrash());
  }

  // ─────────────────────────────────────────────────────────────────────
  //  Maintenance
  //
  //  ancestor_ids และ access_root_id เป็นค่า denormalise ที่เพี้ยนได้
  //  สองตัวนี้มีไว้ตรวจและซ่อม ควรตั้ง cron เรียก /consistency ทุกวันในช่วง
  //  เดือนแรก ๆ แล้ว alert เมื่อไม่เป็นศูนย์
  // ─────────────────────────────────────────────────────────────────────

  @Get('maintenance/consistency')
  async checkConsistency() {
    return unwrap(await this.tree.checkConsistency());
  }

  @Post('maintenance/repair')
  @HttpCode(HttpStatus.OK)
  async repair() {
    return unwrap(await this.tree.repair());
  }

  @Get(':pageId')
  async get(@Param('pageId', ParseUUIDPipe) pageId: string) {
    return unwrap(await this.tree.get(pageId));
  }

  @Post()
  @HttpCode(HttpStatus.CREATED)
  async create(@Body(zodBody(createPageSchema)) body: CreatePageInput) {
    return unwrap(await this.tree.create(body));
  }

  @Patch(':pageId')
  async update(
    @Param('pageId', ParseUUIDPipe) pageId: string,
    @Body(zodBody(updatePageSchema)) body: UpdatePageInput,
  ) {
    return unwrap(await this.tree.update(pageId, body));
  }

  @Post(':pageId/move')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'ย้ายหน้าพร้อมลูกหลานทั้งหมด' })
  async move(
    @Param('pageId', ParseUUIDPipe) pageId: string,
    @Body(zodBody(movePageSchema)) body: MovePageInput,
  ) {
    return unwrap(await this.tree.move(pageId, body));
  }

  @Delete(':pageId')
  @ApiOperation({ summary: 'ย้ายไปถังขยะ (soft delete) พร้อมลูกหลาน' })
  async delete(@Param('pageId', ParseUUIDPipe) pageId: string) {
    return unwrap(await this.tree.delete(pageId));
  }

  @Post(':pageId/restore')
  @HttpCode(HttpStatus.OK)
  async restore(@Param('pageId', ParseUUIDPipe) pageId: string) {
    return unwrap(await this.tree.restore(pageId));
  }

  @Delete(':pageId/purge')
  @ApiOperation({ summary: 'ลบถาวร — owner/admin เท่านั้น ย้อนกลับไม่ได้' })
  async purge(@Param('pageId', ParseUUIDPipe) pageId: string) {
    return unwrap(await this.tree.purge(pageId));
  }
}
